// @aztec/barretenberg/rollup_provider/rollup_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.joinSplitTxFromJson = exports.joinSplitTxToJson = exports.accountTxFromJson = exports.accountTxToJson = exports.pendingTxFromJson = exports.pendingTxToJson = exports.rollupProviderStatusFromJson = exports.rollupProviderStatusToJson = exports.runtimeConfigFromJson = exports.runtimeConfigToJson = exports.txFromJson = exports.txToJson = exports.DefiSettlementTime = exports.TxSettlementTime = void 0;
const blockchain_1 = require("../blockchain");
const bridge_id_1 = require("../bridge_id");
const client_proofs_1 = require("../client_proofs");
const offchain_tx_data_1 = require("../offchain_tx_data");
const tx_id_1 = require("../tx_id");
var TxSettlementTime;
(function (TxSettlementTime) {
    TxSettlementTime[TxSettlementTime["NEXT_ROLLUP"] = 0] = "NEXT_ROLLUP";
    TxSettlementTime[TxSettlementTime["INSTANT"] = 1] = "INSTANT";
})(TxSettlementTime = exports.TxSettlementTime || (exports.TxSettlementTime = {}));
var DefiSettlementTime;
(function (DefiSettlementTime) {
    DefiSettlementTime[DefiSettlementTime["DEADLINE"] = 0] = "DEADLINE";
    DefiSettlementTime[DefiSettlementTime["NEXT_ROLLUP"] = 1] = "NEXT_ROLLUP";
    DefiSettlementTime[DefiSettlementTime["INSTANT"] = 2] = "INSTANT";
})(DefiSettlementTime = exports.DefiSettlementTime || (exports.DefiSettlementTime = {}));
const txToJson = ({ proofData, offchainTxData, depositSignature }) => ({
    proofData: proofData.toString('hex'),
    offchainTxData: offchainTxData.toString('hex'),
    depositSignature: depositSignature ? depositSignature.toString('hex') : undefined,
});
exports.txToJson = txToJson;
const txFromJson = ({ proofData, offchainTxData, depositSignature }) => ({
    proofData: Buffer.from(proofData, 'hex'),
    offchainTxData: Buffer.from(offchainTxData, 'hex'),
    depositSignature: depositSignature ? Buffer.from(depositSignature, 'hex') : undefined,
});
exports.txFromJson = txFromJson;
function runtimeConfigToJson(runtimeConfig) {
    return {
        ...runtimeConfig,
        maxFeeGasPrice: runtimeConfig.maxFeeGasPrice.toString(),
        maxProviderGasPrice: runtimeConfig.maxProviderGasPrice.toString(),
    };
}
exports.runtimeConfigToJson = runtimeConfigToJson;
function runtimeConfigFromJson(runtimeConfig) {
    const { maxFeeGasPrice, maxProviderGasPrice } = runtimeConfig;
    return {
        ...runtimeConfig,
        ...(maxFeeGasPrice !== undefined ? { maxFeeGasPrice: BigInt(maxFeeGasPrice) } : {}),
        ...(maxProviderGasPrice !== undefined ? { maxProviderGasPrice: BigInt(maxProviderGasPrice) } : {}),
    };
}
exports.runtimeConfigFromJson = runtimeConfigFromJson;
function rollupProviderStatusToJson(status) {
    return {
        ...status,
        blockchainStatus: (0, blockchain_1.blockchainStatusToJson)(status.blockchainStatus),
        bridgeStatus: status.bridgeStatus.map(bridge_id_1.bridgeStatusToJson),
        runtimeConfig: runtimeConfigToJson(status.runtimeConfig),
    };
}
exports.rollupProviderStatusToJson = rollupProviderStatusToJson;
function rollupProviderStatusFromJson(status) {
    const { blockchainStatus, nextPublishTime, runtimeConfig, bridgeStatus, ...rest } = status;
    return {
        ...rest,
        blockchainStatus: (0, blockchain_1.blockchainStatusFromJson)(blockchainStatus),
        nextPublishTime: new Date(nextPublishTime),
        runtimeConfig: runtimeConfigFromJson(runtimeConfig),
        bridgeStatus: bridgeStatus.map(bridge_id_1.bridgeStatusFromJson),
    };
}
exports.rollupProviderStatusFromJson = rollupProviderStatusFromJson;
const pendingTxToJson = ({ txId, noteCommitment1, noteCommitment2 }) => ({
    txId: txId.toString(),
    noteCommitment1: noteCommitment1.toString('hex'),
    noteCommitment2: noteCommitment2.toString('hex'),
});
exports.pendingTxToJson = pendingTxToJson;
const pendingTxFromJson = ({ txId, noteCommitment1, noteCommitment2 }) => ({
    txId: tx_id_1.TxId.fromString(txId),
    noteCommitment1: Buffer.from(noteCommitment1, 'hex'),
    noteCommitment2: Buffer.from(noteCommitment2, 'hex'),
});
exports.pendingTxFromJson = pendingTxFromJson;
const accountTxToJson = ({ proofData, offchainTxData }) => ({
    proofData: proofData.proofData.rawProofData.toString('hex'),
    offchainTxData: offchainTxData.toBuffer().toString('hex'),
});
exports.accountTxToJson = accountTxToJson;
const accountTxFromJson = ({ proofData, offchainTxData }) => ({
    proofData: client_proofs_1.AccountProofData.fromBuffer(Buffer.from(proofData, 'hex')),
    offchainTxData: offchain_tx_data_1.OffchainAccountData.fromBuffer(Buffer.from(offchainTxData, 'hex')),
});
exports.accountTxFromJson = accountTxFromJson;
const joinSplitTxToJson = ({ proofData, offchainTxData }) => ({
    proofData: proofData.proofData.rawProofData.toString('hex'),
    offchainTxData: offchainTxData.toBuffer().toString('hex'),
});
exports.joinSplitTxToJson = joinSplitTxToJson;
const joinSplitTxFromJson = ({ proofData, offchainTxData }) => ({
    proofData: client_proofs_1.JoinSplitProofData.fromBuffer(Buffer.from(proofData, 'hex')),
    offchainTxData: offchain_tx_data_1.OffchainJoinSplitData.fromBuffer(Buffer.from(offchainTxData, 'hex')),
});
exports.joinSplitTxFromJson = joinSplitTxFromJson;
