// @aztec/barretenberg/rollup_provider/server_rollup_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerRollupProvider = void 0;
const account_id_1 = require("../account_id");
const asset_1 = require("../asset");
const block_source_1 = require("../block_source");
const iso_fetch_1 = require("../iso_fetch");
const tx_id_1 = require("../tx_id");
const rollup_provider_1 = require("./rollup_provider");
class ServerRollupProvider extends block_source_1.ServerBlockSource {
    constructor(baseUrl, pollInterval = 10000) {
        super(baseUrl, pollInterval);
    }
    async sendTxs(txs) {
        const data = txs.map(rollup_provider_1.txToJson);
        const response = await this.fetch('/txs', data);
        const body = await response.json();
        return body.txIds.map(txId => tx_id_1.TxId.fromString(txId));
    }
    async getTxFees(assetId) {
        const response = await this.fetch('/tx-fees', { assetId });
        const txFees = (await response.json());
        return txFees.map(fees => fees.map(asset_1.assetValueFromJson));
    }
    async getDefiFees(bridgeId) {
        const response = await this.fetch('/defi-fees', { bridgeId: bridgeId.toString() });
        const defiFees = (await response.json());
        return defiFees.map(asset_1.assetValueFromJson);
    }
    async getStatus() {
        const response = await this.fetch('/status');
        try {
            return (0, rollup_provider_1.rollupProviderStatusFromJson)(await response.json());
        }
        catch (err) {
            throw new Error('Bad response: getStatus()');
        }
    }
    async getPendingTxs() {
        const response = await this.fetch('/get-pending-txs');
        const txs = await response.json();
        return txs.map(rollup_provider_1.pendingTxFromJson);
    }
    async getPendingNoteNullifiers() {
        const response = await this.fetch('/get-pending-note-nullifiers');
        const nullifiers = (await response.json());
        return nullifiers.map(n => Buffer.from(n, 'hex'));
    }
    async clientLog(log) {
        await this.fetch('/client-log', log);
    }
    async getInitialWorldState() {
        const response = await this.fetch('/get-initial-world-state');
        const arrBuffer = await response.arrayBuffer();
        return {
            initialAccounts: Buffer.from(arrBuffer),
        };
    }
    async getLatestAccountNonce(accountPubKey) {
        const response = await this.fetch('/get-latest-account-nonce', {
            accountPubKey: accountPubKey.toString(),
        });
        return +(await response.text());
    }
    async getLatestAliasNonce(alias) {
        const response = await this.fetch('/get-latest-alias-nonce', { alias });
        return +(await response.text());
    }
    async getAccountId(alias, nonce) {
        const response = await this.fetch('/get-account-id', { alias, nonce });
        const accountId = await response.text();
        return accountId ? account_id_1.AccountId.fromString(accountId) : undefined;
    }
    async getUnsettledAccountTxs() {
        const response = await this.fetch('/get-unsettled-account-txs');
        const txs = await response.json();
        return txs.map(rollup_provider_1.accountTxFromJson);
    }
    async getUnsettledPaymentTxs() {
        const response = await this.fetch('/get-unsettled-payment-txs');
        const txs = await response.json();
        return txs.map(rollup_provider_1.joinSplitTxFromJson);
    }
    async fetch(path, data) {
        const url = new URL(`${this.baseUrl}${path}`);
        const init = data ? { method: 'POST', body: JSON.stringify(data) } : undefined;
        const response = await (0, iso_fetch_1.fetch)(url.toString(), init).catch(() => undefined);
        if (!response) {
            throw new Error('Failed to contact rollup provider.');
        }
        if (response.status === 400) {
            const body = await response.json();
            throw new Error(body.error);
        }
        if (response.status !== 200) {
            throw new Error(`Bad response code ${response.status}.`);
        }
        return response;
    }
}
exports.ServerRollupProvider = ServerRollupProvider;
