// @aztec/barretenberg/bridge_id/bit_config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BitConfig = void 0;
const getNumber = (val, offset, size) => Number((val >> BigInt(offset)) & ((BigInt(1) << BigInt(size)) - BigInt(1)));
class BitConfig {
    constructor(firstInputVirtual, secondInputVirtual, firstOutputVirtual, secondOutputVirtual, secondInputReal, secondOutputReal) {
        this.firstInputVirtual = firstInputVirtual;
        this.secondInputVirtual = secondInputVirtual;
        this.firstOutputVirtual = firstOutputVirtual;
        this.secondOutputVirtual = secondOutputVirtual;
        this.secondInputReal = secondInputReal;
        this.secondOutputReal = secondOutputReal;
    }
    static fromBigInt(val) {
        return new BitConfig(getNumber(val, 0, 1) == 0 ? false : true, getNumber(val, 1, 1) == 0 ? false : true, getNumber(val, 2, 1) == 0 ? false : true, getNumber(val, 3, 1) == 0 ? false : true, getNumber(val, 4, 1) == 0 ? false : true, getNumber(val, 5, 1) == 0 ? false : true);
    }
    toBigInt() {
        return (BigInt(this.firstInputVirtual) +
            (BigInt(this.secondInputVirtual) << BigInt(1)) +
            (BigInt(this.firstOutputVirtual) << BigInt(2)) +
            (BigInt(this.secondOutputVirtual) << BigInt(3)) +
            (BigInt(this.secondInputReal) << BigInt(4)) +
            (BigInt(this.secondOutputReal) << BigInt(5)));
    }
}
exports.BitConfig = BitConfig;
BitConfig.EMPTY = new BitConfig(false, false, false, false, false, false);
