// @aztec/barretenberg/bridge_id/bridge_id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgeId = void 0;
const asset_1 = require("../asset");
const bigint_buffer_1 = require("../bigint_buffer");
const bit_config_1 = require("./bit_config");
const bridge_id_config_1 = require("./bridge_id_config");
const randomInt = (to = 2 ** 30 - 1) => Math.floor(Math.random() * (to + 1));
const getNumber = (val, offset, size) => Number((val >> BigInt(offset)) & ((BigInt(1) << BigInt(size)) - BigInt(1)));
class BridgeId {
    constructor(addressId, inputAssetIdA, outputAssetIdA, inputAssetIdB, outputAssetIdB, auxData = 0) {
        this.addressId = addressId;
        this.inputAssetIdA = inputAssetIdA;
        this.outputAssetIdA = outputAssetIdA;
        this.inputAssetIdB = inputAssetIdB;
        this.outputAssetIdB = outputAssetIdB;
        this.auxData = auxData;
        const secondInputVirtual = inputAssetIdB !== undefined && (0, asset_1.isVirtualAsset)(inputAssetIdB);
        const secondOutputVirtual = outputAssetIdB !== undefined && (0, asset_1.isVirtualAsset)(outputAssetIdB);
        this.bitConfig = new bit_config_1.BitConfig((0, asset_1.isVirtualAsset)(inputAssetIdA), secondInputVirtual, (0, asset_1.isVirtualAsset)(outputAssetIdA), secondOutputVirtual, inputAssetIdB !== undefined && !secondInputVirtual, outputAssetIdB !== undefined && !secondOutputVirtual);
    }
    static random() {
        return new BridgeId(randomInt(), randomInt(), randomInt(), randomInt(), randomInt(), randomInt());
    }
    static fromBigInt(val) {
        const bitConfig = bit_config_1.BitConfig.fromBigInt(BigInt(getNumber(val, bridge_id_config_1.BITCONFIG_OFFSET, bridge_id_config_1.BITCONFIG_LEN)));
        if (bitConfig.secondInputReal && bitConfig.secondInputVirtual) {
            throw new Error('Invalid second input config.');
        }
        if (bitConfig.secondOutputReal && bitConfig.secondOutputVirtual) {
            throw new Error('Invalid second output config.');
        }
        const hasSecondInput = bitConfig.secondInputReal || bitConfig.secondInputVirtual;
        const hasSecondOutput = bitConfig.secondOutputReal || bitConfig.secondOutputVirtual;
        return new BridgeId(getNumber(val, bridge_id_config_1.ADDRESS_OFFSET, bridge_id_config_1.ADDRESS_BIT_LEN), getNumber(val, bridge_id_config_1.INPUT_ASSET_ID_A_OFFSET, bridge_id_config_1.INPUT_ASSET_ID_A_LEN) +
            (bitConfig.firstInputVirtual ? bridge_id_config_1.virtualAssetIdFlag : 0), getNumber(val, bridge_id_config_1.OUTPUT_ASSET_ID_A_OFFSET, bridge_id_config_1.OUTPUT_ASSET_ID_A_LEN) +
            (bitConfig.firstOutputVirtual ? bridge_id_config_1.virtualAssetIdFlag : 0), hasSecondInput
            ? getNumber(val, bridge_id_config_1.INPUT_ASSET_ID_B_OFFSET, bridge_id_config_1.INPUT_ASSET_ID_B_LEN) +
                (bitConfig.secondInputVirtual ? bridge_id_config_1.virtualAssetIdFlag : 0)
            : undefined, hasSecondOutput
            ? getNumber(val, bridge_id_config_1.OUTPUT_ASSET_ID_B_OFFSET, bridge_id_config_1.OUTPUT_ASSET_ID_B_LEN) +
                (bitConfig.secondOutputVirtual ? bridge_id_config_1.virtualAssetIdFlag : 0)
            : undefined, getNumber(val, bridge_id_config_1.AUX_DATA_OFFSET, bridge_id_config_1.AUX_DATA_LEN));
    }
    static fromBuffer(buf) {
        if (buf.length !== 32) {
            throw new Error('Invalid buffer.');
        }
        return BridgeId.fromBigInt((0, bigint_buffer_1.toBigIntBE)(buf));
    }
    static fromString(str) {
        return BridgeId.fromBuffer(Buffer.from(str.replace(/^0x/i, ''), 'hex'));
    }
    get firstInputVirtual() {
        return this.bitConfig.firstInputVirtual;
    }
    get secondInputVirtual() {
        return this.bitConfig.secondInputVirtual;
    }
    get firstOutputVirtual() {
        return this.bitConfig.firstOutputVirtual;
    }
    get secondOutputVirtual() {
        return this.bitConfig.secondOutputVirtual;
    }
    get secondInputReal() {
        return this.bitConfig.secondInputReal;
    }
    get secondOutputReal() {
        return this.bitConfig.secondOutputReal;
    }
    get numInputAssets() {
        return this.secondInputReal || this.secondInputVirtual ? 2 : 1;
    }
    get numOutputAssets() {
        return this.secondOutputReal || this.secondOutputVirtual ? 2 : 1;
    }
    toBigInt() {
        return (BigInt(this.addressId) +
            (BigInt(this.inputAssetIdA - (this.bitConfig.firstInputVirtual ? bridge_id_config_1.virtualAssetIdFlag : 0)) <<
                BigInt(bridge_id_config_1.INPUT_ASSET_ID_A_OFFSET)) +
            (BigInt((this.inputAssetIdB || 0) - (this.bitConfig.secondInputVirtual ? bridge_id_config_1.virtualAssetIdFlag : 0)) <<
                BigInt(bridge_id_config_1.INPUT_ASSET_ID_B_OFFSET)) +
            (BigInt(this.outputAssetIdA - (this.bitConfig.firstOutputVirtual ? bridge_id_config_1.virtualAssetIdFlag : 0)) <<
                BigInt(bridge_id_config_1.OUTPUT_ASSET_ID_A_OFFSET)) +
            (BigInt((this.outputAssetIdB || 0) - (this.bitConfig.secondOutputVirtual ? bridge_id_config_1.virtualAssetIdFlag : 0)) <<
                BigInt(bridge_id_config_1.OUTPUT_ASSET_ID_B_OFFSET)) +
            (this.bitConfig.toBigInt() << BigInt(bridge_id_config_1.BITCONFIG_OFFSET)) +
            (BigInt(this.auxData) << BigInt(bridge_id_config_1.AUX_DATA_OFFSET)));
    }
    toBuffer() {
        return (0, bigint_buffer_1.toBufferBE)(this.toBigInt(), 32);
    }
    toString() {
        return `0x${this.toBuffer().toString('hex')}`;
    }
    equals(id) {
        return id.toBuffer().equals(this.toBuffer());
    }
}
exports.BridgeId = BridgeId;
BridgeId.ZERO = new BridgeId(0, 0, 0);
BridgeId.ENCODED_LENGTH_IN_BYTES = 32;
