// @aztec/barretenberg/bridge_id/bridge_status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToBridgeStatus = exports.bridgeStatusFromJson = exports.bridgeStatusToJson = void 0;
function bridgeStatusToJson(status) {
    return {
        ...status,
        bridgeId: status.bridgeId.toString(),
    };
}
exports.bridgeStatusToJson = bridgeStatusToJson;
function bridgeStatusFromJson(status) {
    return {
        ...status,
        bridgeId: BigInt(status.bridgeId),
        nextPublishTime: status.nextPublishTime ? new Date(status.nextPublishTime) : undefined,
    };
}
exports.bridgeStatusFromJson = bridgeStatusFromJson;
function convertToBridgeStatus(bridgeConfig, rollupNumber, publishTime, gasAccrued, gasThreshold) {
    return {
        bridgeId: bridgeConfig.bridgeId,
        numTxs: bridgeConfig.numTxs,
        gasThreshold: gasThreshold.toString(),
        gasAccrued: gasAccrued.toString(),
        rollupFrequency: bridgeConfig.rollupFrequency,
        nextRollupNumber: rollupNumber,
        nextPublishTime: publishTime,
    };
}
exports.convertToBridgeStatus = convertToBridgeStatus;
