// @aztec/barretenberg/pippenger/pooled_pippenger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PooledPippenger = void 0;
const tslib_1 = require("tslib");
const single_pippenger_1 = require("./single_pippenger");
const debug_1 = (0, tslib_1.__importDefault)(require("debug"));
const debug = (0, debug_1.default)('bb:pippenger');
class PooledPippenger {
    constructor(workerPool) {
        this.workerPool = workerPool;
        this.pool = [];
    }
    async init(crsData) {
        const start = new Date().getTime();
        debug(`initializing: ${new Date().getTime() - start}ms`);
        this.pool = await Promise.all(this.workerPool.workers.map(async (w) => {
            const p = new single_pippenger_1.SinglePippenger(w);
            await p.init(crsData);
            return p;
        }));
        debug(`initialization took: ${new Date().getTime() - start}ms`);
    }
    async pippengerUnsafe(scalars, from, range) {
        const scalarsPerWorker = range / this.pool.length;
        const start = new Date().getTime();
        const results = await Promise.all(this.pool.map((p, i) => {
            const subset = scalars.slice(scalarsPerWorker * i * 32, scalarsPerWorker * (i + 1) * 32);
            return p.pippengerUnsafe(subset, scalarsPerWorker * i, scalarsPerWorker);
        }));
        debug(`pippenger run took: ${new Date().getTime() - start}ms`);
        return await this.sumElements(Buffer.concat(results));
    }
    async sumElements(buffer) {
        return this.pool[0].sumElements(buffer);
    }
}
exports.PooledPippenger = PooledPippenger;
